Snapshots cards and creation page

torchiaf opened the pull request:
Fixes #4527

### Notes

- `description` field is always empty, see #5580
- dialogs will be implemented in a new PR
- info banners for create/restore/delete actions are implemented in a new PR

### Screenshots

![image](https://github.com/rancher-sandbox/rancher-desktop/assets/26394656/6e44a756-48a9-4a4c-bcd9-63f9ca89eb2a)

![image](https://github.com/rancher-sandbox/rancher-desktop/assets/26394656/c3350e36-7cac-4ab4-99d6-27aa2b95ec92)

![image](https://github.com/rancher-sandbox/rancher-desktop/assets/26394656/5bac654e-22c2-4c98-bd87-6eeed236614c)

torchiaf commented:
## Layout update

![image](https://github.com/rancher-sandbox/rancher-desktop/assets/26394656/b9c75a43-e766-4e58-9509-347e6fb03824)

jandubois commented:
Please hide the UI behind a feature-flag, so we can merge the PR without exposing the UI in the 1.9 release (because the backend changes will not be ready by then).

ericpromislow reviewed (changes requested):
We should impose a limit on how long a snapshot name can be. Maybe 80 characters. or 100. But something...

Ooops ... wrong PR. Please ignore. I can't delete this.

rak-phillip commented on pkg/rancher-desktop/components/SnapshotCard.vue:
```diff
@@ -0,0 +1,128 @@
+  return `${ d.getFullYear() }-${ leadingZero(d.getMonth() + 1) }-${ leadingZero(d.getDate()) } ${ leadingZero(d.getHours()) }:${ leadingZero(d.getMinutes()) }`;
+}
+
+export default Vue.extend<any, any, any, any>({
```
Try to avoid the usage of `any` wherever possible. We should be able to make use of just `Vue.extend` if we provide a type annotation for the `snapshot` computed prop.

```suggestion
export default Vue.extend({
```

rak-phillip commented on pkg/rancher-desktop/components/Snapshots.vue:
```diff
@@ -0,0 +1,61 @@
+import SnapshotCard from '@pkg/components/SnapshotCard.vue';
+
+interface Data {
+  // eslint-disable-next-line no-undef
+  snapshotsPollingInterval: NodeJS.Timer | null,
```
Previously, we've used `ReturnType<typeof setInterval> | undefined` to work with both `setInterval` and `clearInterval` 

```suggestion
  snapshotsPollingInterval: ReturnType<typeof setInterval> | undefined;
```

rak-phillip commented on pkg/rancher-desktop/components/Snapshots.vue:
```diff
@@ -0,0 +1,61 @@
+  snapshotsPollingInterval: NodeJS.Timer | null,
+}
+
+export default Vue.extend<Data, any, any, any>({
```
We can make use of a `VuexBindings` interface to get around relying on `any`

```suggestion
interface VuexBindings {
  snapshots: Snapshot[]
}

export default (Vue as VueConstructor<Vue & VuexBindings>).extend({
```

We have some examples peppered throughout the code. Take a look at for example `pkg/rancher-desktop/pages/Preferences.vue`

rak-phillip commented on pkg/rancher-desktop/pages/snapshots/create.vue:
```diff
@@ -0,0 +1,153 @@
+import Vue from 'vue';
+
+const defaultName = () => {
+  const dateString = new Date()
+    .toISOString()
+    .slice(0, 19)
+    .replace('T', '_')
+    .replace(/-/g, '_')
+    .replace(/:/g, '_');
```
I think that we can use dayjs to do this in a more readable way

```suggestion  
const dateString = dayjs().format('YYYY-MM-DD_HH_mm_ss');
```

rak-phillip commented on pkg/rancher-desktop/pages/snapshots/create.vue:
```diff
@@ -0,0 +1,153 @@
+  creating: boolean,
+}
+
+export default Vue.extend<Data, any, any, any>({
```
Similar to others, we should be able to allow Vue to infer types and remove the usage of `any` if we provide type annotations for the data prop and each of the computed props in this component.

ericpromislow commented:
On Linux, at startup I get the error message Cannot read properties of undefined (reading 'name') in the main window, with no routing menu. The dev console coimplains that /Snapshots isn't found in the routes array, that `custom validator check faile dof rprop "items" (in components/Nav.vue), and that the <undefined.name> is in some Vue/Proxy list-renderer code (Nav.vue is the only app-specific file in the stacktrace ).

On macOS, yarn dev just hangs. I was able to build a packaged app, and it starts up, but with the same main-window as on Linux (and because it's a packaged build there's no console)

Also there are nuxt-related merge-conflicts. I merged them anyway (just by deleting files), but wonder if that nuxt-change broke this branch, because something's working in those screenshots.

ericpromislow reviewed (changes requested):
Doesn't work on macOS nor linux

rak-phillip reviewed (changes requested):
This is looking quite good overall. I think there are a few things that we can clean up on this first pass.

torchiaf commented on pkg/rancher-desktop/pages/Snapshots.vue:
Yeah, we discussed about this part with QA, I'll add the message as we do in Diagnostic page.

torchiaf commented:
Thanks for the review!
@rak-phillip I changed some small thing about Vue typing, It should be correct now but we can always fix these small details at the end of the whole implementation.
@ericpromislow Do you have probablye some problems in rebasing the master with new routing files? I pushed the rebase and it works for me.

rak-phillip commented:
> Thanks for the review! @rak-phillip I changed some small thing about Vue typing, It should be correct now but we can always fix these small details at the end of the whole implementation. Also added the no-rows template to the main page:
> 
> ![image](https://user-images.githubusercontent.com/26394656/273248782-a1dcef37-c148-4519-8d26-7e1ecb270324.png)

Thanks, I think that will be better. In hindsight, I think that including default text in the was a mistake. I created #5673 to address this.

torchiaf commented:
> > Thanks for the review! @rak-phillip I changed some small thing about Vue typing, It should be correct now but we can always fix these small details at the end of the whole implementation. Also added the no-rows template to the main page:
> > ![image](https://user-images.githubusercontent.com/26394656/273248782-a1dcef37-c148-4519-8d26-7e1ecb270324.png)
> 
> Thanks, I think that will be better. In hindsight, I think that including default text for each slot was a mistake. I created #5673 to address this.

The screenshots is wrong, I've fixed the body message i my code before pushing:

https://github.com/rancher-sandbox/rancher-desktop/blob/b9834f7ad5a2c79744310bab54b44aa262541cfe/pkg/rancher-desktop/components/Snapshots.vue#L80

rak-phillip commented on pkg/rancher-desktop/components/Snapshots.vue:
```diff
@@ -0,0 +1,85 @@
+
+  watch: {
+    snapshots: {
+      handler(neu) {
```
I know this is a bit of a nitpick, but can we use a better name than `neu`? I can't say that I ever really understood what `neu` means and we just as easily name this something along the lines of `val` or `newVal`.

torchiaf replied:
😅 I think 'neu' is new in german, this is used since 'new' would be a reserved word.

rak-phillip commented on pkg/rancher-desktop/components/Snapshots.vue:
```diff
@@ -0,0 +1,85 @@
+
+  computed: { ...mapGetters('snapshots', { snapshots: 'list' }) },
+
+  watch: {
+    snapshots: {
+      handler(neu) {
+        this.isEmpty = neu?.length === 0;
+        this.$emit('change', neu);
+      },
+    },
```
I think that this watcher would better fit as an `isEmpty` computed prop. Is there a reason why a watcher would be required here?

rak-phillip commented on pkg/rancher-desktop/pages/PortForwarding.vue:
```diff
@@ -154,6 +154,6 @@ export default Vue.extend({
 <style scoped>
   .content {
-    padding: 20px;
+    padding-top: 13px;
```
Why are we changing the Port Forwarding page here?

torchiaf replied:
This is a small issue in the Forwarding page layout that I found and fixed here.

rak-phillip commented on pkg/rancher-desktop/pages/snapshots/create.vue:
```diff
@@ -0,0 +1,158 @@
+      'page/setHeader',
+      { title: this.t('snapshots.create.title') },
+    );
+    (this.$refs.nameInput as any)?.focus();
```
Is the `v-focus` directive not functioning properly?

torchiaf replied:
This can be removed.

rak-phillip commented on pkg/rancher-desktop/store/snapshots.ts:
```diff
@@ -0,0 +1,74 @@
+  async fetch({ commit, rootState }: SnapshotsActionContext) {
+    const response = await fetchAPI('/v1/snapshots', rootState);
+
+    if (!response.ok) {
+      console.log(`fetchSnapshots: failed: status: ${ response.status }:${ response.statusText }`);
+
+      return;
+    }
```
Do we need to relay that there was an error in the UI or will this be in a follow-up PR?

torchiaf replied:
This is required in a follow-up, we will need to notify errors during creation and restore to the user.

rak-phillip reviewed (changes requested):
I think that I have questions more than anything at this point. We should sync early next week on how we want to handle `EmptyContent`

torchiaf commented on pkg/rancher-desktop/components/SnapshotCard.vue:
```diff
@@ -0,0 +1,139 @@
+  value: Snapshot
+}
+
+export default Vue.extend<Computed, Methods, Computed, Props>({
```
There are other cases in the UI where the Vue typing is wrong:
`DiagnosticBody`
![image](https://github.com/rancher-sandbox/rancher-desktop/assets/26394656/b6c436b2-2fba-48aa-ba85-6244bf06304d)
I would open a separated issue for that.

rak-phillip replied:
As a rule, we have cleaned these up when we going back to make changes to a problematic component. But you are correct, an issue would be the proper way to track these so I created #5686. Please add to the list if you think we're missing any TypeScript components.

torchiaf commented on pkg/rancher-desktop/components/Snapshots.vue:
```diff
@@ -0,0 +1,85 @@
+    snapshots: {
+      handler(neu) {
+        this.isEmpty = neu?.length === 0;
+        this.$emit('change', neu);
```
Yeh, we can remove it. It was needed in the first design but now it's useless.

rak-phillip reviewed (approved):
LGTM

ericpromislow reviewed (changes requested):
We need a way to let users sort snapshots, preferably by both name and creation time, and allow ascending and descending. Don't we have a widget to make this nearly trivial?

jandubois commented:
> We need a way to let users sort snapshots, preferably by both name and creation time, and allow ascending and descending.

This is not a requirement for the 1.11 release; please create a new issue for 1.12.

For 1.11 it is sufficient to display in reverse creation order (newest on top).

@ericpromislow Please remove your "Request changes" block, if this is the only part of the PR that you are objecting to.
